roachtest: tpccbench/nodes=6/cpu=16/multi-az failed

roachtest.tpccbench/nodes=6/cpu=16/multi-az [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2920165&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2920165&tab=artifacts#/tpccbench/nodes=6/cpu=16/multi-az) on master @ [3cd29547884feb24b150b67ca2a88073798a35fd](https://github.com/cockroachdb/cockroach/commits/3cd29547884feb24b150b67ca2a88073798a35fd):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/tpccbench/nodes=6/cpu=16/multi-az/run_1
	cluster.go:1716,context.go:140,cluster.go:1705,test_runner.go:858: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-2920165-1619291008-63-n7cpu16-geo --oneshot --ignore-empty-nodes: exit status 1 7: skipped
		1: dead
		3: 1587
		2: 1572
		4: 1588
		6: 1605
		5: 1598
		Error: UNCLASSIFIED_PROBLEM: 1: dead
		(1) UNCLASSIFIED_PROBLEM
		Wraps: (2) attached stack trace
		  -- stack trace:
		  | main.glob..func14
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:1162
		  | main.wrap.func1
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:284
		  | github.com/spf13/cobra.(*Command).execute
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:856
		  | github.com/spf13/cobra.(*Command).ExecuteC
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:960
		  | github.com/spf13/cobra.(*Command).Execute
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:897
		  | main.main
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:2057
		  | runtime.main
		  | 	/usr/local/go/src/runtime/proc.go:204
		  | runtime.goexit
		  | 	/usr/local/go/src/runtime/asm_amd64.s:1374
		Wraps: (3) 1: dead
		Error types: (1) errors.Unclassified (2) *withstack.withStack (3) *errutil.leafError
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
# From https://go.crdb.dev/p/roachstress, perhaps edited lightly.
caffeinate ./roachstress.sh tpccbench/nodes=6/cpu=16/multi-az
```
</p>
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #62339 roachtest: tpccbench/nodes=6/cpu=16/multi-az failed [overload] [C-test-failure O-roachtest O-robot branch-release-21.1]
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*tpccbench/nodes=6/cpu=16/multi-az.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

This test was actually a "success", but it left the cluster in an unexpected state since overloaded VMs prevented `roachprod stop`:

```
03:41:00 cluster.go:387: > /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod stop teamcity-2920165-1619291008-63-n7cpu16-geo:1-6
teamcity-2920165-1619291008-63-n7cpu16-geo: stopping and waiting...............
0: exit status 255: 
1: exit status 255: 
2: exit status 255: 
3: exit status 255: 
4: exit status 255: 
5: exit status 255: 
I210425 03:41:16.273227 1 (gostd) cluster_synced.go:1649  [-] 1  command failed
03:41:16 tpcc.go:824: unable to stop cluster; retrying to allow vm to recover: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod stop teamcity-2920165-1619291008-63-n7cpu16-geo:1-6 returned: exit status 1
03:41:46 cluster.go:1169: test status: stopping nodes :1-6
03:41:46 cluster.go:387: > /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod stop teamcity-2920165-1619291008-63-n7cpu16-geo:1-6
teamcity-2920165-1619291008-63-n7cpu16-geo: stopping and waiting.
03:41:48 cluster.go:1169: test status: starting nodes :1-6
03:41:48 cluster.go:387: > /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod start --encrypt=false teamcity-2920165-1619291008-63-n7cpu16-geo:1-6
teamcity-2920165-1619291008-63-n7cpu16-geo: starting nodes
teamcity-2920165-1619291008-63-n7cpu16-geo: initializing cluster
teamcity-2920165-1619291008-63-n7cpu16-geo: setting cluster settings
03:42:22 tpcc.go:899: test status: running benchmark, warehouses=3750
03:42:22 cluster.go:2404: [w2] > ./cockroach workload run tpcc --warehouses=5000 --active-warehouses=3750 --tolerate-errors --ramp=5m0s --duration=10m0s --histograms=perf/warehouses=3750/stats.json {pgurl:1-6}
2: 1572
3: 1587
4: 1588
1: 1570
6: 1605
5: 1598
1: kill exited nonzero
1: dead
03:54:39 tpcc.go:976: --- SEARCH ITER FAIL: TPCC 3750 resulted in 0.0 tpmC and failed due to monitor failure: unexpected node event: 1: dead
03:54:39 tpcc.go:985: ------
MAX WAREHOUSES = 3737

...

teardown: 03:55:58 test_runner.go:646: [w2] --- FAIL: tpccbench/nodes=6/cpu=16/multi-az (14118.46s)
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/tpccbench/nodes=6/cpu=16/multi-az/run_1
	cluster.go:1716,context.go:140,cluster.go:1705,test_runner.go:858: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-2920165-1619291008-63-n7cpu16-geo --oneshot --ignore-empty-nodes: exit status 1 7: skipped
		1: dead
		3: 1587
		2: 1572
		4: 1588
		6: 1605
		5: 1598
		Error: UNCLASSIFIED_PROBLEM: 1: dead
```

We have an implicit assumption that the CRDB cluster is running at the end of the line search. Here we hit a case in which the last step in the line search was a failure due to a dead node, which should not have failed the test but it did anyway. Will send a patch.

--- a/pkg/cmd/roachtest/tpcc.go
+++ b/pkg/cmd/roachtest/tpcc.go
@@ -793,18 +793,8 @@ func runTPCCBench(ctx context.Context, t *test, c *cluster, b tpccBenchSpec) {
 		t.Fatal(errors.Wrap(err, "failed to create temp dir"))
 	}
 	defer func() { _ = os.RemoveAll(resultsDir) }()
-	s := search.NewLineSearcher(1, b.LoadWarehouses, b.EstimatedMax, initStepSize, precision)
-	iteration := 0
-	if res, err := s.Search(func(warehouses int) (bool, error) {
-		iteration++
-		t.l.Printf("initializing cluster for %d warehouses (search attempt: %d)", warehouses, iteration)
 
-		// NB: for goroutines in this monitor, handle errors via `t.Fatal` to
-		// *abort* the line search and whole tpccbench run. Return the errors
-		// to indicate that the specific warehouse count failed, but that the
-		// line search ought to continue.
-		m := newMonitor(ctx, c, roachNodes)
-
+	restart := func() {
 		// We overload the clusters in tpccbench, which can lead to transient infra
 		// failures. These are a) really annoying to debug and b) hide the actual
 		// passing warehouse count, making the line search sensitive to the choice
@@ -841,12 +831,28 @@ func runTPCCBench(ctx context.Context, t *test, c *cluster, b tpccBenchSpec) {
 		}
 
 		c.Start(ctx, t, append(b.startOpts(), roachNodes)...)
+	}
+
+	s := search.NewLineSearcher(1, b.LoadWarehouses, b.EstimatedMax, initStepSize, precision)
+	iteration := 0
+	if res, err := s.Search(func(warehouses int) (bool, error) {
+		iteration++
+		t.l.Printf("initializing cluster for %d warehouses (search attempt: %d)", warehouses, iteration)
+
+		restart()
+
 		time.Sleep(restartWait)
 
 		// Set up the load generation configuration.
 		rampDur := 5 * time.Minute
 		loadDur := 10 * time.Minute
 		loadDone := make(chan time.Time, numLoadGroups)
+
+		// NB: for goroutines in this monitor, handle errors via `t.Fatal` to
+		// *abort* the line search and whole tpccbench run. Return the errors
+		// to indicate that the specific warehouse count failed, but that the
+		// line search ought to continue.
+		m := newMonitor(ctx, c, roachNodes)
 
 		// If we're running chaos in this configuration, modify this config.
 		if b.Chaos {
@@ -981,6 +987,10 @@ func runTPCCBench(ctx context.Context, t *test, c *cluster, b tpccBenchSpec) {
 	}); err != nil {
 		t.Fatal(err)
 	} else {
+		// The last iteration may have been a failing run that overloaded
+		// nodes to the point of them crashing. Make roachtest happy by
+		// restarting the cluster so that it can run consistency checks.
+		restart()
 		ttycolor.Stdout(ttycolor.Green)
 		t.l.Printf("------\nMAX WAREHOUSES = %d\n------\n\n", res)
 		ttycolor.Stdout(ttycolor.Reset)

